Resources/Computing/SingularityComputingElement.py

```diff
@@ -154,7 +154,11 @@ def __hasSingularity(self):
 
   @staticmethod
   def __findInstallBaseDir():
-    return os.readlink(os.path.join(DIRAC.rootPath, "bashrc"))
+    """Find the path to root of the current DIRAC installation"""
+    candidate = os.path.join(DIRAC.rootPath, "bashrc")
+    if not os.path.exists(candidate):
+      raise NotImplementedError("Failed to find a candidate bashrc")
```

Comment:
This exception would not be handled by the caller.

Reply:
I'm not quite sure what you mean by this, are you suggesting:
```diff
-    if not os.path.exists(candidate):
-      raise NotImplementedError("Failed to find a candidate bashrc")
```

This exception cannot be handled as it means something is fundamentally wrong with this bit of code and the current installation layout. I think it's easier to be able to grep for `Failed to find a candidate bashrc` instead of getting a weird error message about files not being found somewhere later in the code.

Reply:
I mean that we never raise exceptions, but rather return `S_ERROR` (this is a policy in DIRAC). When you call `__findInstallBaseDir` (2 times in this module) you should do it inside a `try..except` and return `S_ERROR`.

I might also (personally) find the raising of exceptions easier to code/handle (as a code flow) but this is just not the policy, historically.

Reply:
As long as it is done properly, consistently, in a closed context,  and carefully reviewed, I think we should review that policy :-) Maybe it is not the case here, because it would require to modify to much code above. But sticking to old policy just because of principles is not really a way forward. Anyway, not a discussion for here, but I would advocate that we do not object to exception based on this old policy

Reply:
I'm just going to remove the `NotImplementedError` and if it's a problem for someone's installation layout it can raise an exception from somewhere deep down in the standard library instead.

Reply:
> As long as it is done properly, consistently, in a closed context, and carefully reviewed, 

Yes, in a closed context. There are examples of code in DIRAC where exceptions are defined, raised, and handled, e.g. see the Workflow package modules. But at the end only `S_OK` and `S_ERROR` are returned.

> I think we should review that policy :-) 

I am not against the exceptions: I coded the Workflow modules and I used there quite some exceptions. Regarding the policy: see below.

> sticking to old policy just because of principles is not really a way forward. Anyway, not a discussion for here, but I would advocate that we do not object to exception based on this old policy

I think I didn't express myself correctly: this principle is there because, in a _general_ case, if we introduce widely exceptions, then we would need to start considering that **any** method could return the following:
- `S_OK`
- `S_ERROR`
- an exception

While the first 2 are "known" by the client code (they are the implicit API), the third "not necessarily" (do I need to expand this point?).


Now, for the specific case of this PR, the only "known" caller of `SingularityComputingElement` is this code: https://github.com/DIRACGrid/DIRAC/blob/integration/Resources/Computing/ComputingElementFactory.py#L29 which do handle possible exceptions, so for this case it would be _probably_ OK to leave it. 





ps: I linked this thread in https://indico.cern.ch/event/977355/ (next BiLD agenda)

Reply:
> I'm just going to remove the NotImplementedError and if it's a problem for someone's installation layout it can raise an exception from somewhere deep down in the standard library instead.

No that's not I meant to do.

Reply:
At the moment this code will never raise an exception unless there is actually a bug in the code. If there are logic errors then they will be handled in exactly the same way as any other in DIRAC. There is an interesting longer-term discussion to be had for the BiLD but I think this PR is finished.